### PR TITLE
Implement PersistentIndex merge compaction

### LINKED ISSUE #3228: Implement MutableIndex and ImmutableIndex merge compaction

### PR DESCRIPTION
**Pull request by decster**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [ ] enhancement
- [x] others

## Which issues of this PR fixes ：

Fixes #3228

## Problem Summary(Required) ：
This PR implements PersistentIndex merge compaction, which merges l0 and l1 into a new persistent index l1.
This merge process first estimate the number of shards of the new l1 index, then merge each shard  sequentially

### CODE REVIEW

File: be/src/storage/persistent_index.cpp [hunk 1]
```diff
@@ -1176,4 +1257,137 @@ Status PersistentIndex::_delete_expired_index_file(const EditVersion& version) {
...
+    //         cur_shard_idx:    0     1     2     3
+    for (size_t l1_shard_idx = 0; l1_shard_idx < nshard_l1; l1_shard_idx++) {
+        RETURN_IF_ERROR(_l1->_get_kvs_for_shard(l1_kvs_by_shard, l1_shard_idx));
+        size_t num_shard_finished = (l1_shard_idx + 1) * nshard / nshard_l1;
+        std::vector<KVRef> kvs;
+        while (cur_shard_idx < num_shard_finished) {
+            kvs.clear();
+            RETURN_IF_ERROR(merge_shard_kvs(_key_size, l0_kvs_by_shard[cur_shard_idx], l1_kvs_by_shard[cur_shard_idx],
+                                            estimated_size_per_shard, kvs));
+            RETURN_IF_ERROR(writer.write_shard(_key_size, value_size, npage_hint, kvs));
```
[the remark this answers is not in the document]
Comment (decster) [reply be/src/storage/persistent_index.cpp:1381]: Writer destructor

[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/storage/persistent_index.cpp:1381]: If crashed after failed?

File: be/src/storage/persistent_index.cpp [hunk 2]
```diff
@@ -416,52 +423,93 @@ StatusOr<std::unique_ptr<ImmutableIndexShard>> ImmutableIndexShard::create(size_
...
+            Env::Default()->delete_file(_idx_file_path_tmp);
+        }
+    }
+
+    Status init(const string& dir, const EditVersion& version) {
+        _version = version;
+        _idx_file_path = strings::Substitute("$0/index.l1.$1.$2", dir, version.major(), version.minor());
+        _idx_file_path_tmp = _idx_file_path + ".tmp";
+        fs::BlockManager* block_mgr = fs::fs_util::block_manager();
+        fs::CreateBlockOptions wblock_opts({_idx_file_path_tmp});
```
Comment (sevev) [inline be/src/storage/persistent_index.cpp:439]: The default mode of CreateBlockOptions is `MUST_CREATE`. If be crash during merge compaction and tmp file is not deleted, the next merge compaction maybe failed.

File: be/src/storage/persistent_index.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (decster) [reply be/src/storage/persistent_index.cpp:1381]: Then only sweep thread can cleanup this, currently, there are many index file names not checked by sweep thread, will add this logic in future PR.

**Review by chaoyli — APPROVED** [review]

**Review by Astralidea — APPROVED** [review]

**Comment by Astralidea** [comment]
run starrocks_clang-format

**Comment by chaoyli** [comment]
run starrocks_clang-format

**Comment by chaoyli** [comment]
run starrocks_clang-format

**Pull request merged**